src/Channels.Networking.TLS/SecurityContext.cs

```diff
@@ -0,0 +1,198 @@
+                        _initOkay = true;
+                        GetCredentials();
+                        return;
+                    }
+                }
+                throw new InvalidOperationException($"Unable to find the security package named {SecurityPackage}");
+            }
+            finally
+            {
+                if (secPointer != null)
+                {
+                    InteropSspi.FreeContextBuffer((IntPtr)secPointer);
+                }
+            }
+        }
+
+        private unsafe void GetCredentials()
+        {
+            CredentialUse direction;
+            CredentialFlags flags;
+            if (_isServer)
+            {
+                direction = CredentialUse.Inbound;
+                flags = CredentialFlags.UseStrongCrypto | CredentialFlags.SendAuxRecord;
+            }
+            else
+            {
+                direction = CredentialUse.Outbound;
+                flags = CredentialFlags.ValidateManual | CredentialFlags.NoDefaultCred | CredentialFlags.SendAuxRecord | CredentialFlags.UseStrongCrypto;
+            }
+
+            var creds = new SecureCredential()
+            {
+                rootStore = IntPtr.Zero,
+                phMappers = IntPtr.Zero,
+                palgSupportedAlgs = IntPtr.Zero,
+                cMappers = 0,
+                cSupportedAlgs = 0,
+                dwSessionLifespan = 0,
+                reserved = 0,
+                dwMinimumCipherStrength = 0, //this is required to force encryption
+                dwMaximumCipherStrength = 0,
+                version = SecureCredential.CurrentVersion,
+                dwFlags = flags,
+                certContextArray = IntPtr.Zero,
+                cCreds = 0
+            };
+
+            IntPtr certPointer;
+            if (_isServer)
+            {
+                creds.grbitEnabledProtocols = InteropSspi.ServerProtocolMask;
+                certPointer = _serverCertificate.Handle;
+                //pointer to the pointer
+                IntPtr certPointerPointer = new IntPtr(&certPointer);
+                creds.certContextArray = certPointerPointer;
+                creds.cCreds = 1;
+            }
+            else
+            {
+                creds.grbitEnabledProtocols = InteropSspi.ClientProtocolMask;
+            }
+
+            long timestamp = 0;
+            SecurityStatus code = (SecurityStatus)InteropSspi.AcquireCredentialsHandleW(null, SecurityPackage, (int)direction
+                , null, ref creds, null, null, ref _credsHandle, out timestamp);
+
+            if (code != SecurityStatus.OK)
+            {
+                throw new InvalidOperationException($"Could not acquire the credentials with return code {code}");
+            }
+        }
+
+        public SecureChannel CreateSecureChannel(IChannel channel)
+        {
+            var chan = new SecureChannel(channel, _channelFactory);
+            chan.StartReading(new SecureConnectionContext(this));
```

Comment:
Reduced performance. if you wait until the connection point before deciding if it is a client or server connection, you have to call the API to setup an Authentication context for each connection. This way we can reuse the same authentication context over and over as long as you are using the same settings/certs.